src/runtime_src/hip/core/memory_pool.cpp

```diff
@@ -0,0 +1,407 @@
+      *reinterpret_cast<uint64_t*>(value) = m_release_threshold;
+      break;
+
+    case hipMemPoolAttrReservedMemCurrent:
+      *reinterpret_cast<uint64_t*>(value) = m_reserved_mem_current;
+      break;
+
+    case hipMemPoolAttrReservedMemHigh:
+      *reinterpret_cast<uint64_t*>(value) = m_reserved_mem_high;
+      break;
+    
+    case hipMemPoolAttrUsedMemCurrent:
+      *reinterpret_cast<uint64_t*>(value) = m_used_mem_current;
+      break;
+
+    case hipMemPoolAttrUsedMemHigh:
+      *reinterpret_cast<uint64_t*>(value) = m_used_mem_high;
+      break;
+    };
+  }
+
+  void
+  memory_pool::set_attribute(hipMemPoolAttr attr, void* value)
+  {
+    if (m_list.size() == 0) {
+      init();
+    }
+
+    switch (attr)
+    {
+    case hipMemPoolReuseFollowEventDependencies:
+      m_reuse_follow_event_dependencies = *reinterpret_cast<int*>(value);
+      break;
+
+    case hipMemPoolReuseAllowOpportunistic:
+      m_reuse_allow_opportunistic = *reinterpret_cast<int*>(value);
+      break;
+
+    case hipMemPoolReuseAllowInternalDependencies:
+      m_reuse_allow_internal_dependencies = *reinterpret_cast<int*>(value);
+      break;
+
+    case hipMemPoolAttrReleaseThreshold:
+      m_release_threshold = *reinterpret_cast<uint64_t*>(value);
+      break;
+
+    case hipMemPoolAttrReservedMemCurrent:
+      m_reserved_mem_current = *reinterpret_cast<uint64_t*>(value);
+      break;
+
+    case hipMemPoolAttrReservedMemHigh:
+      m_reserved_mem_high = *reinterpret_cast<uint64_t*>(value);
+      break;
+
+    case hipMemPoolAttrUsedMemCurrent:
+      m_used_mem_current = *reinterpret_cast<uint64_t*>(value);
+      break;
+
+    case hipMemPoolAttrUsedMemHigh:
+      m_used_mem_high = *reinterpret_cast<uint64_t*>(value);
+      break;
+    };
+  }
+
+  bool
+  memory_pool::extend_memory_list(size_t size)
+  {
+    m_list.insert(m_list.begin(), std::make_shared<memory_pool_node>(m_device, size, m_last_id++));
+    return true;
+  }
+
+  void memory_pool::malloc(void**ptr, size_t size)
+  {
+    *ptr = alloc(size);
+  }
+
+  void*
+  memory_pool::alloc(size_t size)
+  {
+    if (m_list.size() == 0) {
+      init();
+    }
+
+    size_t page_size = xrt_core::getpagesize();
+    size_t aligned_size = ((size + page_size) / page_size) * page_size;
+
+    std::lock_guard lock(m_mutex);
+
+    if (aligned_size > m_pool_size)
+    {
+      throw std::runtime_error("requested size is greater than memory pool block size.");
+    }
+
+    std::shared_ptr<memory_pool_node> mm;
+
+    uint32_t num_tries = 0;
+    // find first free slot that fits, if none is found, enlarge the pool and try one more time
+    while (num_tries <= 2) {
+      ++ num_tries;
+      auto mp_itr = m_list.begin();
+      while (mp_itr != m_list.end())
+      {
+        mm = *mp_itr;
+        if (m_pool_size - mm->m_used < aligned_size)
+        {
+          mp_itr++;
+          continue;
+        }
+
+        auto _free = mm->m_free_list;
+        auto _not_free = _free;
```

Comment:
Please never leading underscores.  Plus `not_free` makes no sense as it is assigned `free` ???

Reply:
underscores leading has been removed and variable name was changed from "not_free" to "alloc_slot", since it represent the slot that is chosen from current free slots for allocation.